### DISCUSSION TITLE
typescript like instance variables

### DISCUSSION THREAD

**Discussion by HakamFostok** [dotnet/csharplang#5059] [Language Ideas] [answered]
# Typescript like instance variables

## Summary
The use of the typescript-syntax to define instance variables in the class.

## Motivation
ASP.NET Core default to use Dependency-Injection was a great step.
After this (correct and very elegant, at least in my opinion) decision, the adoption of DI has increased because now it is the default in the ASP.NET Core.
Unfortunately, there is too much boilerplate to implement simple injection in the class

```
public class Service
{
     private readonly IServiceA _serviceA;
     private readonly IServiceB _serviceB;

     public Service(IServiceA serviceA, IServiceB serviceB)
     {
           _serviceA = serviceA;
           _serviceB = serviceB;
     }
}
```

I think DI is here to stay, and it achieves one of the SOLID principles
so having C# language support for DI will make the adoption and use of the DI more prevalent and easier.

## Detailed design

The design is very simple here is an example

```
public class Service
{
     public Service(
         private readonly IServiceA serviceA, 
         private readonly IServiceB serviceB)
     {
     }
}
```

## Drawbacks
I am aware of the issues that can be caused because of the constructor overloading.
Because of that, the suggestion to resolve these issues is to prevent constructor overloading completely when this syntax is used.
When using syntax the goal is to make DI easier, It could be suitable for other cases also but, of course, it is not a sliver-bullet.
In some cases, this syntax is a perfect fit and in others, it is not. 
It is just an option to remove a lot of boilerplate and ceremony when DI is needed.

## Further details
- This feature is not related or intended to enhance the record syntax or anything else, which is discussed in other issues.
- At the beginning, this could be supported only on classes. `records` and `struct` can come later, if ever.
- All access modifiers that are allowed on the instance variable in the class could be allowed in this syntax.
- The `readonly` modifier is optional, it could be omitted.
- There is no need to support `const` or `static` keywords for those types of instance variables.
- The code in the (only) constructor can be executed after setting all the variables.
- The use of the `base()` syntax to call the base constructor should be the same, and it will start before the setting of the instance variables.

**Comment by HaloFour** [comment, marked as answer]
See: https://github.com/dotnet/csharplang/issues/2691

**Comment by TonyValenti** [comment]
Once Required Properties gets implemented, I might never write another constructor again.  Required Properties seems like a much better alternative to constructors.

**Reply by HakamFostok** [reply]
The default DI used by .NET does not support Property Injection it supports only Constructor Injection, I am talking specifically about DI support by C# language, I do not think this issue is related to a required properties, which is a great feature but can not substitute constructors in 100% of the cases.

**Reply by jnm2** [reply]
I expect DI containers will support required properties pretty fast after they ship. I know of at least one that plans to.
